### DISCUSSION TITLE
[Suggestions] Hero screens, enhancement to "click and hold and move" the mouse cursor, "drag, collect and drop"

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#4883] [Ideas]
I suggest to allow to "drag, collect and drop" for an even more flexible use of the "click and hold and move".

In fheroes2:

Army troop positions at the start.
![Drag collect drop 001](https://user-images.githubusercontent.com/43583994/147878969-2d20a18f-b0da-4d26-b07d-3edd222d2caf.png)

"click and hold and move" the mouse cursor to an empty slot to move the origin troop.
![Drag collect drop flèche 01 001](https://user-images.githubusercontent.com/43583994/147878976-08401364-d29f-4fab-bf37-2a55787ceb8b.png)
![Drag collect drop 002](https://user-images.githubusercontent.com/43583994/147879004-4c118c85-7851-441a-9fd4-f1202fb7f360.png)

"click and hold and move" the mouse cursor to a troop of another type to swap these two troops.
![Drag collect drop flèche 02 001](https://user-images.githubusercontent.com/43583994/147879063-347d6f5f-f15c-4903-a442-378665c00584.png)
![Drag collect drop 003](https://user-images.githubusercontent.com/43583994/147879059-f92b041e-5586-4c88-9aa7-66c95da24c5d.png)

"click and hold and move" the mouse cursor to a troop of the same type (without flying over a troop of the same type) to unite these two troops.
![Drag collect drop flèche 03 001](https://user-images.githubusercontent.com/43583994/147879164-d8d1567d-22ba-4a8b-9d42-5ce147e91a2f.png)
![Drag collect drop 004](https://user-images.githubusercontent.com/43583994/147879168-03301182-cf3a-4ab9-b1cb-59b74ac4d778.png)

"click and hold and move" the mouse cursor to a troop of the same type (by flying over one or more troops of the same type) to collect and unite all these troops of the same type.
![Drag collect drop flèche 04 001](https://user-images.githubusercontent.com/43583994/147879215-c0a62471-3bf0-4e6a-a506-08c2211b5e43.png)
![Drag collect drop 005](https://user-images.githubusercontent.com/43583994/147879211-7002762e-9f8a-4c46-a55f-4effd21fde25.png)

Army troop positions at the start.
![Drag collect drop 006](https://user-images.githubusercontent.com/43583994/147879312-b564ff54-9181-4378-a4ea-c0a20807f1d1.png)

"click and hold and move" the mouse cursor to a troop of the same type (by flying over one or more troops of the same type) to collect and unite all these troops of the same type.
First exemple.
![Drag collect drop flèche 01 008](https://user-images.githubusercontent.com/43583994/147879361-5afa503b-c708-49dc-8a4d-47e097a0d136.png)
![Drag collect drop 008](https://user-images.githubusercontent.com/43583994/147879365-59d1a651-3c77-4cd9-aa07-0afc4c21ee14.png)

"click and hold and move" the mouse cursor to a troop of the same type (by flying over one or more troops of the same type) to collect and unite all these troops of the same type.
Second exemple.
![Drag collect drop flèche 02 006](https://user-images.githubusercontent.com/43583994/147879413-53712bd8-99c9-4e0b-b2ec-f273ed38b7de.png)
![Drag collect drop 007](https://user-images.githubusercontent.com/43583994/147879408-5fd73236-ac80-434c-96ac-8dd89c031599.png)

**Edit 2022/01/03.**

Related issue: https://github.com/ihhub/fheroes2/issues/4850.
Related PR: https://github.com/ihhub/fheroes2/pull/4855.

**Comment by Branikolog** [comment]
Hi, @LeHerosInconnu.
Actually the behaviour of "Click-hold and move" is what I was talking in the #4855. The logic here should drop similar, as we have in traditional clicking operation: we should be able to drop target troop at any place in hero's army initiating "move", "unite" and "switch", depending on a slot we're placing target troop (empty slot/same troop in other slot/other type of troop in busy slot respectively).

Second part of your proposal, called "Drag-collect-drop" is quite interesting, but requires some practice and skills. Me actually find it a bit tricky in a case we have 3 stacks, but the task is to unite the most right and the most left stacks without uniting with middle one. To implement such behaviour I suggest to not initiate "collecting" while simple dragging over army slots.
In a case we want to initiate "collecting" I suggest to proceed the above movements with pressed ALT button and "flying" over needed stacks. One more suggestion from my side here: while "flying" over same type stacks we can show red selection frame on the stacks, which are going to be united till the moment we release left mouse button. In a case we want to cancel "collecting", we need to release ALT button.

Anyway, I find this logic quite complicated and implement "drag-and-drop" (click, hold and move) logic first.

**Reply by LeHerosInconnu** [reply]
Hello @Branikolog,

> Hi, @LeHerosInconnu. Actually the behaviour of "Click-hold and move" is what I was talking in the #4855. The logic here should drop similar, as we have in traditional clicking operation: we should be able to drop target troop at any place in hero's army initiating "move", "unite" and "switch", depending on a slot we're placing target troop (empty slot/same troop in other slot/other type of troop in busy slot respectively).

Yes, it's a reminder of the possibilities mentioned.

> Second part of your proposal, called "Drag-collect-drop" is quite interesting, but requires some practice and skills.
> Me actually find it a bit tricky in a case we have 3 stacks, but the task is to unite the most right and the most left stacks without uniting with middle one.

In this case it is better to use the traditional way of doing a left click on the starting troop and a left click on the destination troop. :)

> To implement such behaviour I suggest to not initiate "collecting" while simple dragging over army slots. In a case we want to initiate "collecting" I suggest to proceed the above movements with pressed ALT button and "flying" over needed stacks.

No, the idea is to use only the mouse, so without using the keyboard, otherwise there is no point, and you might as well use the traditional way of doing a combination of left clicks.

> One more suggestion from my side here: while "flying" over same type stacks we can show red selection frame on the stacks, which are going to be united till the moment we release left mouse button. In a case we want to cancel "collecting", we need to release ALT button.

Showing the red selection for the troops that will be reunited is a good idea. :)

> Anyway, I find this logic quite complicated and implement "drag-and-drop" (click, hold and move) logic first.

Everything is in the same logic according to the destination slots, free slot (move), slot with a troop of different type (swap), slot with a troop of the same type (reunite), slot with a troop of the same type by flying over one or more troops of the same type (collect and reunite).

**Reply by Branikolog** [reply]
>No, the idea is to use only the mouse, so without using the keyboard, otherwise there is no point, and you might as well use the traditional way of doing a combination of left clicks.

Aren't we talking here about just an alternative behaviour of arranging army without single clicks, but with dragging technique? I suppose, we offer two similar ways of interacting for those players, who got used to another way of moving UI elements in "drag-and-drop" way, don't we? So "collecting" is an new way of rearranging the army (which offers not such a common use), so it could be used with combination with some of keyboard buttons. 
Btw, we can implement same "collecting" technique for traditional "clicking, while we select 1 stack, press ALT and then select all other needed stacks to be united together. So in this way these both click and dragging techniques (As I described before) will work in similar way. So player can choose, which is more comfortable for him: drag or click.
Lets be honest: we can't offer all possible ways of interaction with mouse only. And I know the majority of players often use keyboard, since it's much faster than using mouse only. For mouse only mode we've already added the most common scenarios (fast split, making 1 unit as default split offer for pressing MIN button). So adding Shift/Alt/Ctrl for drag-and-drop interaction will make both ways of mouse interaction similar, so players can choose, which one is more comfortable for him.

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> > No, the idea is to use only the mouse, so without using the keyboard, otherwise there is no point, and you might as well use the traditional way of doing a combination of left clicks.
> 
> Aren't we talking here about just an alternative behaviour of arranging army without single clicks, but with dragging technique? I suppose, we offer two similar ways of interacting for those players, who got used to another way of moving UI elements in "drag-and-drop" way, don't we? So "collecting" is an new way of rearranging the army (which offers not such a common use), so it could be used with combination with some of keyboard buttons. Btw, we can implement same "collecting" technique for traditional "clicking, while we select 1 stack, press ALT and then select all other needed stacks to be united together. So in this way these both click and dragging techniques (As I described before) will work in similar way. So player can choose, which is more comfortable for him: drag or click. Lets be honest: we can't offer all possible ways of interaction with mouse only. And I know the majority of players often use keyboard, since it's much faster than using mouse only. For mouse only mode we've already added the most common scenarios (fast split, making 1 unit as default split offer for pressing MIN button). So adding Shift/Alt/Ctrl for drag-and-drop interaction will make both ways of mouse interaction similar, so players can choose, which one is more comfortable for him.

In these proposals (mine), the goal is to use only the mouse.
The mouse is already in the user's hand, he doesn't need to take his eyes off the screen or use his other hand to operate the keyboard.

After that, in general, it should be possible to add ways of doing things using the keyboard, but this should be an alternative to using the mouse, not a combined use of the mouse and the keyboard, and I don't see the need for it here.

In the cases presented it is possible to do it with only the mouse.

**Reply by felix642** [reply]
I'll have to agree with @Branikolog on this one.
I'd rather use drag-collect and drop in combination with a key since I don't feel like this new implementation is a common use. 
If I need to combine troops I usually don't need to collect the ones in my path and I don't really like this to be a default behavior. 
I think it would instead confuse players since they expect drag and drop to be the default behavior. 
We can't implement everything using only the mouse since there are too many options (troop stats, move troop, combine troops, split troops, swap troops) and we don't have to choice but to default to the usage of keys. 

The one proposed where :
Drag & drop = move / combine troops
Shift + drag & drop = split troops
Alt + drag & drop = drag + collect and drop

Sounds like a pretty good idea to me.

**Reply by LeHerosInconnu** [reply]
Hello @felix642 and @Branikolog,

> I'll have to agree with @Branikolog on this one. I'd rather use drag-collect and drop in combination with a key since I don't feel like this new implementation is a common use. If I need to combine troops I usually don't need to collect the ones in my path and I don't really like this to be a default behavior. I think it would instead confuse players since they expect drag and drop to be the default behavior. We can't implement everything using only the mouse since there are too many options (troop stats, move troop, combine troops, split troops, swap troops) and we don't have to choice but to default to the usage of keys.
> 
> The one proposed where : Drag & drop = move / combine troops Shift + drag & drop = split troops Alt + drag & drop = drag + collect and drop
> 
> Sounds like a pretty good idea to me.

Since it is already possible to use the left mouse click for move and combine operations, why not reserve drag and drop only for collecting troops.
It's a bit wasteful to use the mouse for two similar actions when it could be better used (and there is no need to use the keyboard).

Left click (selection) plus left click (destination) to move or join.
Left click (selection) plus right click (destination) to separate or rearrange.
Drag and drop is only used to join hovered troops of the same type.

It then becomes possible to go further and split a troop into several troops using the same principle.
Left click and hold on a troop, move the cursor over empty slots (from one empty slot up to four empty slots), and the starting troop is divided into the free slots hovered.

So, from the starting troop, hovering and finishing the mouse drag on a troop of the same type unites all the troops of the same type hovered.
And, from the starting troop, hovering and finishing the mouse drag on an empty slot distributes the starting troop to all empty slots hovered (similar to the fast separation into slots window operation, but without opening the window).
